unison-cli/src/Unison/Share/Sync.hs

```diff
@@ -0,0 +1,764 @@
+            False -> pure (Left (FastForwardPushErrorNoWritePermission repoPath))
+            True ->
+              doFastForwardPath (localHeadHash : localTailHashes) <&> \case
+                Share.FastForwardPathSuccess -> Right ()
+                Share.FastForwardPathMissingDependencies (Share.NeedDependencies dependencies) ->
+                  Left (FastForwardPushErrorServerMissingDependencies dependencies)
+                -- Weird: someone must have force-pushed no history here, or something. We observed a history at this
+                -- path but moments ago!
+                Share.FastForwardPathNoHistory -> Left (FastForwardPushErrorNoHistory repoPath)
+                Share.FastForwardPathNoWritePermission _ -> Left (FastForwardPushErrorNoWritePermission repoPath)
+                Share.FastForwardPathNotFastForward _ -> Left FastForwardPushErrorNotFastForward
+  where
+    doUpload :: List.NonEmpty CausalHash -> IO Bool
+    -- Maybe we could save round trips here by including the tail (or the head *and* the tail) as "extra hashes", but we
+    -- don't have that API yet. So, we only upload the head causal entity (which we don't even know for sure the server
+    -- doesn't have yet), and will (eventually) end up uploading the casuals in the tail that the server needs.
+    doUpload (headHash :| _tailHashes) =
+      uploadEntities
+        httpClient
+        unisonShareUrl
+        conn
+        (Share.RepoPath.repoName repoPath)
+        (NESet.singleton (causalHashToHash headHash))
+
+    doFastForwardPath :: [CausalHash] -> IO Share.FastForwardPathResponse
+    doFastForwardPath causalSpine =
+      httpFastForwardPath
+        httpClient
+        unisonShareUrl
+        Share.FastForwardPathRequest
+          { hashes = map causalHashToHash causalSpine,
+            path = repoPath
+          }
+
+    -- Return a list from newest to oldest of the ancestors between (excluding) the latest local and the current remote hash.
+    -- note: seems like we /should/ cut this short, with another command to go longer? :grimace:
+    fancyBfs :: CausalHash -> Share.Hash -> Sqlite.Transaction (Maybe [CausalHash])
+    fancyBfs h0 h1 =
+      tweak <$> dagbfs (== Share.toBase32Hex h1) Q.loadCausalParentsByHash (Hash.toBase32Hex (unCausalHash h0))
+      where
+        -- Drop 1 and reverse (under a Maybe, and twddling hash types):
+        --
+        --   tweak [] = []
+        --   tweak [C,B,A] = [A,B]
+        --
+        -- The drop 1 is because dagbfs returns the goal at the head of the returned list, but we know what the goal is
+        -- already (the remote head hash).
+        tweak :: Maybe [Base32Hex] -> Maybe [CausalHash]
+        tweak =
+          fmap (map (CausalHash . Hash.fromBase32Hex) . reverse . drop 1)
+
+data Step a
+  = DeadEnd
+  | KeepSearching (List.NonEmpty a)
+  | FoundGoal a
+
+-- FIXME: document
+dagbfs :: forall a m. Monad m => (a -> Bool) -> (a -> m [a]) -> a -> m (Maybe [a])
```

Comment:
Not necessarily any better or worse, but I got nerd-sniped by BFS, so here's a version of the same which uses logict;

```haskell
dagbfs :: forall a m. (MonadIO m, Show a) => (a -> Bool) -> (a -> m [a]) -> a -> m (Maybe [a])
dagbfs goal getChildren root =
  fmap listToMaybe . observeManyT 1 . flip runReaderT [] $ (bfs root >>= guardMaybe)
  where
    guardMaybe Nothing = empty
    guardMaybe (Just a) = pure a
    bfs :: a -> ReaderT [a] (LogicT m) (Maybe [a])
    bfs child = local (child :) $ do
      liftIO $ print child
      -- For a proper breadth-first traversal we emit failures too.
      let nodeResult =
            if goal child
              then Just <$> ask
              else pure Nothing
      nodeResult <|> do
        children <- lift . lift $ getChildren child
        interleaveAll (bfs <$> children)

-- |
-- >>> observeAll $ interleaveAll [pure @Logic 1, pure 2 <|> pure 3 <|> pure 4, pure 5, pure 6 <|> pure 7]
-- [1,2,5,6,3,7,4]
interleaveAll :: forall f a. MonadLogic f => [f a] -> f a
interleaveAll [] = empty
interleaveAll xs = do
  go xs []
  where
    go :: [f a] -> [f a] -> f a
    go [] ms = interleaveAll ms
    go (x : rest) ms = do
      msplit x >>= \case
        Nothing -> go rest ms
        Just (a, m) -> pure a <|> go rest (ms ++ [m])
```